### ISSUE TITLE
Omit removes rest object props

### ISSUE THREAD

**Issue by vladshcherbin** [fabian-hiller/valibot#481]
Rest object props with `unknown()` are removed by `omit()`

Example:

```js
import { object, omit, parse, string, unknown } from 'valibot'

const data = {
  a: 'a',
  b: 'b',
  c: 'c',
  d: 'd'
}

const schema = omit(
  object({
    a: string(),
    b: string()
  }, unknown()),
  ['a']
)

console.log(parse(schema, data))
```

## Current output

```js
{
  b: 'b'
}
```

## Expected output

```js
{
  b: 'b',
  c: 'c',
  d: 'd'
}
```

valibot 0.30.0

**Comment by fabian-hiller** [comment]
This is the intended behavior. It is documented in our [API reference](https://valibot.dev/api/omit/#explanation), but I can understand why this might be confusing. I will think about it for further development of the library.

**Comment by vladshcherbin** [comment]
@fabian-hiller I've read the note after opening the issue, still didn't figure out the `must be added again` part to make expected output. Used `transform` as a workaround for now.

If it's somehow possible to make it work using `must be added again` (adding `unknown()` again or smth else) I'd be happy to see an example and add to the docs 🙌

**Comment by fabian-hiller** [comment]
Check out our [playground](https://valibot.dev/playground/?code=JYWwDg9gTgLgBAKjgQwM5wG5wGZQiOAcg2QBtgAjCGQgbgCh6BjCAO1XgHkKArAUyYwAykwAWfEMgCMcALyYAdBF4CYACnpw4Ab01a4Aaz4BPKQC5FHKMFYBzNQEoANHq1HjAJgsYFVm-edXQxMAZm9fGGs7RxctAF9YxQBXVgNWCAB3Vkd6BwZmNg44bn5BEXFJDzlFfGB1EtVyiWknOABtQncPQgBdVp8UtMzshzzGFnYIUj4FUgh7PR9UZGw+AAVkKFQ+NQaysWaPVu1g0wtCQlb3MKJLuGpxKHPCODiHXNogA) for an example.

```ts
import * as v from 'valibot';

const ObjectSchema1 = v.object(
  {
    key1: v.string(),
    key2: v.string(),
    key3: v.string(),
  },
  v.unknown()
);

const ObjectSchema2 = v.omit(
  ObjectSchema1,
  ['key2'],
  v.unknown() // <-- Add `rest` again
);
```

**Comment by vladshcherbin** [comment]
@fabian-hiller thank you for the example. I'm still confused 😅

Here's my [playground](https://valibot.dev/playground/?code=JYWwDg9gTgLgBAKjgQwM5wG5wGZQiOAcg2QBtgAjCGQgbgCh6BjCAO1XgHkKArAUyYwAykwAWfEMgCMcALyYAdBF4CYACnpw4Ab01a4Aaz4BPKQC5FHKMFYBzNQEoANHq1HjAJgsYFVm-edXQxMAZm9fGGs7RxctAF9YxQBXVgNWCAB3Vkd6BwZmNg44bn5BEXFJDzlFfGB1EtVyiWknOABtQncPQgBdVp8UtMzshzzGFnZ4ABNkGGRq7WDTC06pQlaulYNujdCtkPW4anEoFYhCODjxwvgScimANTIkvmqfMGQoVD41BrKxZoeVozOYOa7sCCkPgKUgQex3YCPZ58MYTVCQ6Gw+FkRFPUgvBRdBxAA)

From what I expect, `key2` should be omitted.
However it's still in the output.
Basically `ObjectSchema2` does nothing 🤔

This key is also removed from autocomplete but we see it's there in the output:

<img width="798" alt="image" src="https://github.com/fabian-hiller/valibot/assets/6711845/83f9ac7c-ece6-44fd-a813-4b183d485e22">


I remember trying it with the same result.

p.s. it may be the intended behavior and just not the one I'd expect from the code

**Comment by fabian-hiller** [comment]
Because of the `rest` argument `v.unknown()` the schema allow any unknown entry to pass.

**Comment by vladshcherbin** [comment]
@fabian-hiller yeah, it kinda reverts what `omit()` does.

So basically it's not possible to omit `key2` from object and leave rest unknown props untouched using `omit()` function, only `transform` can do this.

I'd consider this a bug in `omit()` since it clearly does more than just omitting `key2` by omitting all unknown props from the object.

**Comment by fabian-hiller** [comment]
How would you do the same with pure TypeScript?

**Comment by vladshcherbin** [comment]
@fabian-hiller a quick one:

```ts
const data = {
  a: 'a',
  b: 'b',
  c: 'c',
  d: 'd'
}

function omit<T, K extends keyof T>(object: T, keys: K[]): Omit<T, K> {
  const result = { ...object }

  keys.forEach((key) => {
    delete result[key]
  })

  return result
}

const result = omit(data, ['a', 'b'])

console.log(result)
console.log(result.c)
```

I'd probably use some kind of `Object.keys`, `Object.entries`, `Object.fromEntries` or `reduce` but it requires a lot more time spent on typings 😅

**Comment by fabian-hiller** [comment]
And how would you implement it if you wanted it to allow unknown entries to pass? I am not sure if I understand you correctly, but I think you want to obmit a specific key, but also allow any unknown key to pass. If that's your goal, there's probably only one solution: You need to explicitly declare that key as `never`.

**Comment by vladshcherbin** [comment]
Yes, thats exactly what I want - omit only specific key and leave the rest untouched (be it specified keys from schema or unknown ones).

```ts
// initial object
const data = { a: 1, b: 2, c: 3, d: 4 }

// remove ONLY a
// result is { b: 2, c: 3, d: 4 }
omit(data, ['a'])

// remove ONLY a AND c
// result is { b: 2, d: 4 }
omit(data, ['a', 'c'])
```

## incorrect

I've tried Typescript [Omit](https://www.typescriptlang.org/docs/handbook/utility-types.html#omittype-keys) for correct output type but it's not smart enough for index signature:

```ts
type Data = {
  [key: string]: unknown
  a: number
  b: number
  c: number
  d: number
}

type Omitted = Omit<Data, 'a'>
```

<img width="610" alt="image" src="https://github.com/fabian-hiller/valibot/assets/6711845/e2f1d609-68cb-43ac-938b-7814e8dd23ef">


The output type loses the overview of defined keys.

## correct

The correct one can be seen using `Except` from magical 🧙 [type-fest](https://github.com/sindresorhus/type-fest?tab=readme-ov-file#utilities):

```ts
import type { Except } from 'type-fest'

type Data = {
  [key: string]: unknown
  a: number
  b: number
  c: number
  d: number
}

type Omitted = Except<Data, 'a'>
```

<img width="620" alt="image" src="https://github.com/fabian-hiller/valibot/assets/6711845/fa065a41-ad19-4129-842d-4557c6b696d4">

No wonder [type-fest](https://www.npmjs.com/package/type-fest) has **over 160 million** downloads 😅

That's basically what I expect `omit` to do.

I also believe this is how other libraries work (tested on [lodash omit](https://lodash.com/docs/4.17.15#omit))

- TS issue: https://github.com/microsoft/TypeScript/issues/45367 (other can exist too)
- Zod issue with suggested workaround: https://github.com/colinhacks/zod/issues/2052 (hello `transform`)

**Comment by vladshcherbin** [comment]
Same `bug` conclusion from `type-fest` with identical examples and expectations - https://github.com/sindresorhus/type-fest/issues/382

**Comment by fabian-hiller** [comment]
Thank you very much! I will look into this in the next few weeks. Maybe we should also add an `except` method in addition to the `omit` functionality.

**Comment by ct-gdf** [comment]
Hey folks! I have been trying out the latest (`0.31.0-rc.6`) and was noticing some other breaking changes regarding `.omit` with the new pipeline changes. Can repro as a separate issue, but I was able to repro in the playground. At the same time, I was able to find a workaround. Maybe helpful in the docs?

**Comment by fabian-hiller** [comment]
This is documented here: https://valibot.dev/api/omit/

> Because `omit` changes the data type of the input and output, it is not allowed to pass a schema that has been modified by the `pipe` method, as this may cause runtime errors. Please use the `pipe` method after you have modified the schema with `omit`.

I do not recommend your workaround, instead write your schema this way:

```ts
import * as v from "valibot";

const Schema = v.object({
  email: v.pipe(v.string(), v.email()),
  password: v.pipe(v.string(), v.minLength(8)),
});

const Without = v.pipe(v.omit(Schema, ["password"]), v.brand("Hello"));
```

**Comment by ct-gdf** [comment]
Thank you for the reply! 
I missed that callout in the docs. A code example would make this more apparent I think.

fwiw, my real use case had a base Schema that included a brand `A` that I was omitting a single property from to define the new Schema. Are you saying that the more correct approach would be to let the base Schema be defined unbranded, then derive the two variants with their own distinct brand?

**Comment by fabian-hiller** [comment]
> Are you saying that the more correct approach would be to let the base Schema be defined unbranded, then derive the two variants with their own distinct brand?

Yes, this is the recommended way at this time. I understand that this can be cumbersome in some cases. That's why I will consider adding something like a `removePipe` method. That way you could simply remove the pipe before applying `omit`. Here is an example:

```ts
import * as v from "valibot";

const Schema = v.pipe(
  v.object({
    email: v.pipe(v.string(), v.email()),
    password: v.pipe(v.string(), v.minLength(8)),
  }),
  v.brand("Hello"),
);

const Without = v.omit(v.removePipe(Schema), ["password"]);
```

What do you think of this idea? What could be alternative names for this function? I prefer short but still meaningful names.

**Comment by ct-gdf** [comment]
I do not have a strong preference either way. This was the only case where I _thought_ I needed something like `removePipe`.

**Comment by fabian-hiller** [comment]
Okay. Then let's wait for more feedback from other developers to make a good decision in the long run.

**Comment by fabian-hiller** [comment]
Yes, `omit` is implemented similarly to TypeScript's `Omit` type, and I don't think we should change that. The problem is that the specified `rest` contains every key, including `key1` and `key3`. This feature will probably require us to implement a new `except` method.

**Comment by fabian-hiller** [comment]
Have a look at the source code in line 407. As written in the docs, `omit` just creates a copy and removes the specified entries: https://github.com/fabian-hiller/valibot/blob/main/library/src/methods/omit/omit.ts

**Comment by vladshcherbin** [comment]
@fabian-hiller yes, I also expect `omit` to _remove_ the entries

In [the latest example from above](https://valibot.dev/playground/?code=JYWwDg9gTgLgBAKjgQwM5wG5wGZQiOAcg2QBtgAjCGQgKFoGMIA7VeAeRGBhgFMATAMoMAFrxDI4AXkwA6fNwAUtOHIgUAVrwYwA6txEAlXm0UBvFargBrXgE8AjAC45bKMGYBzRQEoANJaqtnYATC4Ysm4e3v6BNvYAzOGRMO5evgGqAL5+cgCuzNbMEADuzL6xqgDahMEOhLm1iYQAurQ+ANxwAPTdcCB5bHAUvHBm8aEuUV5wWfRMrPD8yDCSMhZB9s5EyA2WwWFEFHubdklEDHRzjCxDUCZ5pPAyEWDIUKi8ipzcfEKi4mQuWWqx881uEFIvFkpAg3nuqEeMB8QA) I have input data:

```ts
{
  key1: 'a',
  key2: 'b',
  key3: 'c'
}
```

now omit:

```ts
omit(object, ['key1', 'key3'])
``` 

here I expect `omit` to remove specified `key1` and `key3`, but it actually does nothing for`objectWithRest`, the result object **is not changed**.

that's why I'd consider it as a bug or a very unreliable method and at least add a note about this behavior in the docs - personally I find the outcome result surprising

**Comment by vladshcherbin** [comment]
This is actually not the same issue I mentioned when I opened this issue - it works correctly now in latest version.

A new issue should be probably opened for tracking `objectWithRest` + `omit` combination and other devs awareness.
